server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -905,7 +909,7 @@ private static Mapper getMapper(final ParseContext context, ObjectMapper objectM
         }
         String leafName = subfields[subfields.length - 1];
         mapper = objectMapper.getMapper(leafName);
-        if (mapper != null) {
+        if (mapper != null || isObject) {
```

Comment:
nit: should we instead split the method into two, one for objects and one for leaf fields. The one for leaves applies also runtime fields at the end, while the object variant doesn't?

Reply:
I did think about that, but it ends up repeating a lot of fairly complex code so I think it's better kept like this for now

Reply:
ok, maybe I am missing something then, I thought that one method (for the leaves) could call the other (for objects) and also apply runtime fields if it gets null. What would cause duplication?

Reply:
oh, yes, that would make sense.  Have pushed a change!